### PR TITLE
feat(hyd): Refactor the THS to work with both mechanical and electrical control

### PR DESCRIPTION
**Pull request by lukecologne**
## Summary of Changes

This PR refactors the existing THS code so that it works with both mechanical control input (i.e. trim wheel + electric motors) like in the A320, and direct electric commands from the computers, like in the A380.

## Additional context

Discord username (if different from GitHub):

## Testing instructions

Make sure the THS works (both manual and automatic control, including manual override when in automatic control, and blocking of the trim wheel with green/yellow hyd lost)

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

**Review by 2hwk — APPROVED** [review]

**Comment by 2hwk** [comment]
Testing instructions?

**Comment by lukecologne** [comment]
Done

File: fbw-a380x/src/wasm/systems/a380_systems/src/hydraulic/mod.rs [hunk 1]
```diff
@@ -5663,6 +5660,113 @@ impl SimulationElement for ElevatorSystemHydraulicController {
...
+        self.update_ths_controllers_solenoids();
+    }
+
+    fn update_ths_controllers_positions(&mut self) {
+        self.controllers[ThsMotorPosition::Green as usize].set_requested_position(
+            self.position_requests_from_fbw[ThsMotorPosition::Green as usize],
+        );
+        self.controllers[ThsMotorPosition::Yellow as usize].set_requested_position(
+            self.position_requests_from_fbw[ThsMotorPosition::Yellow as usize],
+        );
```
Comment (Gurgel100) [inline fbw-a380x/src/wasm/systems/a380_systems/src/hydraulic/mod.rs:5747]:
```suggestion
        for (controller, position_request) in self
            .controllers
            .iter_mut()
            .zip(self.position_requests_from_fbw)
        {
            controller.set_requested_position(position_request);
        }
```

File: fbw-a380x/src/wasm/systems/a380_systems/src/hydraulic/mod.rs [hunk 2]
```diff
@@ -5663,6 +5660,113 @@ impl SimulationElement for ElevatorSystemHydraulicController {
...
+    }
+
+    fn update_ths_controllers_solenoids(&mut self) {
+        self.controllers[ThsMotorPosition::Green as usize]
+            .set_mode(self.solenoid_energized_from_fbw[ThsMotorPosition::Green as usize]);
+        self.controllers[ThsMotorPosition::Yellow as usize]
+            .set_mode(self.solenoid_energized_from_fbw[ThsMotorPosition::Yellow as usize]);
+    }
```
Comment (Gurgel100) [inline fbw-a380x/src/wasm/systems/a380_systems/src/hydraulic/mod.rs:5755]:
```suggestion
        for (controller, solenoid_energized) in self
            .controllers
            .iter_mut()
            .zip(self.solenoid_energized_from_fbw)
        {
            controller.set_mode(solenoid_energized);
        }
```

File: fbw-common/src/wasm/systems/systems/src/hydraulic/trimmable_horizontal_stabilizer.rs [hunk 3]
```diff
@@ -263,9 +272,10 @@ impl HydraulicDriveMotor {
     }
 
     fn current_max_speed_from_hydraulic_pressure(&self, pressure: Pressure) -> AngularVelocity {
-        let pressure_coefficient = (pressure.get::<psi>() / Self::MAX_SPEED_HYD_PRESSURE_PSI)
-            .min(1.)
-            .max(0.);
+        let pressure_coefficient = (pressure.get::<psi>()
+            / self.max_speed_hyd_pressure.get::<psi>())
+        .min(1.)
+        .max(0.);
```
Comment (Gurgel100) [inline fbw-common/src/wasm/systems/systems/src/hydraulic/trimmable_horizontal_stabilizer.rs:278]:
```suggestion
        let pressure_coefficient =
            (pressure.get::<psi>() / self.max_speed_hyd_pressure.get::<psi>()).clamp(0., 1.);
```

File: fbw-common/src/wasm/systems/systems/src/hydraulic/trimmable_horizontal_stabilizer.rs [hunk 4]
```diff
@@ -514,10 +524,44 @@ impl SimulationElement for PitchTrimActuator {
...
+struct ThsMotorMechanicalController {
+    should_motor_activate: bool,
+    requested_position: Angle,
+}
+impl ThsMotorMechanicalController {
+    fn new() -> Self {
+        Self {
+            should_motor_activate: false,
+
+            requested_position: Angle::new::<degree>(0.),
```
Comment (Gurgel100) [inline fbw-common/src/wasm/systems/systems/src/hydraulic/trimmable_horizontal_stabilizer.rs:537]:
```suggestion
            requested_position: Angle::default(),
```

File: fbw-common/src/wasm/systems/systems/src/hydraulic/trimmable_horizontal_stabilizer.rs [hunk 5]
```diff
@@ -573,11 +628,20 @@ impl TrimmableHorizontalStabilizerAssembly {
...
+            controller.update(
+                pressures[controller_index],
+                self.ths_deflection_range
+                    * self
+                        .pitch_trim_actuator
+                        .position_normalized()
+                        .get::<ratio>()
+                    + self.min_ths_deflection,
+            );
+        }
```
Comment (Gurgel100) [inline fbw-common/src/wasm/systems/systems/src/hydraulic/trimmable_horizontal_stabilizer.rs:641]:
```suggestion
        for (controller, pressure) in self.controllers.iter_mut().zip(pressures) {
            controller.update(
                pressure,
                self.ths_deflection_range
                    * self
                        .pitch_trim_actuator
                        .position_normalized()
                        .get::<ratio>()
                    + self.min_ths_deflection,
            );
        }
```

File: fbw-common/src/wasm/systems/systems/src/hydraulic/trimmable_horizontal_stabilizer.rs [hunk 6]
```diff
@@ -656,22 +729,22 @@ impl TrimmableHorizontalStabilizerHydraulics {
...
-            * trim_actuator_output_position_normalized.get::<ratio>()
-            + self.min_deflection;
-
-        self.update_spool_valve_lock_position(deflection_demand);
+        // Both controllers send the same position command for the A320, which is the aircraft that will use the
+        // spool valve lock position. The A380 will not use it, since it doesn't have a manual trim wheel.
+        // So, we just use the first controller position.
+        self.update_spool_valve_lock_position(controllers[0].requested_position());
 
         for (motor_index, motor) in self.hydraulic_motors.iter_mut().enumerate() {
```
Comment (Gurgel100) [inline fbw-common/src/wasm/systems/systems/src/hydraulic/trimmable_horizontal_stabilizer.rs:740]: I know the code was already there before but nevertheless:
```suggestion
        for ((motor, controller), pressure) in self
            .hydraulic_motors
            .iter_mut()
            .zip(controllers)
            .zip(pressures)
        {
            motor.update(
                context,
                controller.motor_should_activate(),
                self.actual_deflection,
                controller.requested_position(),
                pressure,
            );
        }
```
Unfortunately I can't select multiple lines in this instance.

**Review by Gurgel100 — CHANGES_REQUESTED** [review]
In this PR you are introducing a naming mixture. The old struct are named `TrimmableHorizontalStabilizerActuator` for example but newly introduced structs are named `ThsSystemHydraulicController` for example. See https://github.com/flybywiresim/a32nx/blob/master/fbw-common/src/wasm/systems/guidelines.md#naming for reference.
But I can see why using `Ths` instead of the full name as it gets really long.

File: fbw-a380x/src/wasm/systems/a380_systems/src/hydraulic/mod.rs (hunk 1 above)
Comment (lukecologne) [reply fbw-a380x/src/wasm/systems/a380_systems/src/hydraulic/mod.rs:5747]: I think in these cases I'd rather explicitly go through the array, asthere are only two controllers, the looped variant is longer, and also IIRC in the rest of the file, the position requests are all explicitly listed too

**Comment by lukecologne** [comment]
Thanks for the review! Never noticed that I mixed abbreviated and fully name, I'll probably use the full name so I only have to rename the new structs but the full name is indeed a bit long

**Comment by lukecologne** [comment]
Alright, I think that should be everything renamed

File: fbw-a380x/src/wasm/systems/a380_systems/src/hydraulic/mod.rs [hunk 7]
```diff
@@ -5991,6 +6099,11 @@ enum ElevatorPanelPosition {
     Inward = 1,
 }
 
+enum ThsMotorPosition {
```
Comment (Gurgel100) [inline fbw-a380x/src/wasm/systems/a380_systems/src/hydraulic/mod.rs:6102]: You missed one ;)
```suggestion
enum TrimmableHorizontalStabilizerMotorPosition {
```

**Review by Gurgel100 — APPROVED** [review]

**Comment by 2hwk** [comment]
Quality Assurance Report

Discord: 2Cas#1022
Object of testing: https://github.com/flybywiresim/a32nx/pull/7717
Tier of Testing : 1
Date : 12/02/2023

Testing Process:
Peripherals Tested - Flight Stick Trim Switch, Throttle Trim Wheel, Mouse (click and drag), Mouse (mousewheel), Keyboard
Loaded into KFBW.
-[x] Moved THS with no hyd - Blocked
-[x] Started APU/ENG 2 - Moved THS
-[x] Turned off Yellow HYD Pump - Blocked THS
- [x] Turned on Yellow ELEC pump - Moved THS
- [x] Turned off ELEC Pump, turned on yellow HYD Pump - Moved THS
- [x] Started Flight as per normal
- [x] Held THS in place after takeoff
- [x] Released
- [x] Repeated THS tests on ground in mid air
- [x] Loaded in mid-air
- [x] Restarted mid-air tests   
Negatives:
N/A

Testing Results:
Passed

Conclusions:
LGTM

**Pull request merged**